src/execution/mod.rs

```diff
@@ -218,22 +233,35 @@ pub(crate) fn run_main(
 ///     flow: Dataflow to run.
 ///
 ///     input_builder: Returns input that each worker thread should
-///         process.
+///         process. If you are recovering a stateful dataflow, your
+///         code should resume from the last finalized epoch.
 ///
 ///     output_builder: Returns a callback function for each worker
 ///         thread, called with `(epoch, item)` whenever and item
 ///         passes by a capture operator on this process.
-///
+///         
 ///     addresses: List of host/port addresses for all processes in
 ///         this cluster (including this one).
 ///
 ///     proc_id: Index of this process in cluster; starts from 0.
 ///
+///     recovery_config: State recovery config. See
+///         `bytewax.recovery`. If `None`, state will not be persisted.
+///
 ///     worker_count_per_proc: Number of worker threads to start on
 ///         each process.
-#[pyfunction]
+#[pyfunction(
+    flow,
+    input_builder,
+    output_builder,
+    addresses,
+    proc_id,
+    "*",
```

Comment:
what's this little asterisk bud telling me?

Reply:
This is saying the remainder of the arguments are "keyword only" arguments which require the call site to name each argument. Thus it would be a... `TypeError` (I think) to call `cluster_main(flow, ib, ob, address, proc_id, recovery_store)`. You have to pass all arguments after the `*` by keyword `cluster_main(flow, ib, ob, address, proc_id, recovery_store=recovery_store)`.

The general idea is everything after the `*` is optional and non-hierarchical (e.g. specifying `worker_count_per_proc` doesn't imply that you should be setting `recovery-store`), so you need to pass it by keyword to make it so you don't have to repeat defaults for things you don't care about.

I couldn't find a concise summary of the different parameter passing "modes" but https://peps.python.org/pep-0570/ and https://peps.python.org/pep-3102/ discuss the details of `*` and `/`. Relatedly there's arg and kwarg unpacking with `*list_of_args`, and `**dict_of_kwargs`.

Reply:
Thank you. I googled "asterisk python arguments" and fainted at the number of results